java/team25core/FourWheelDirectIMUDrivetrain.java

```diff
@@ -0,0 +1,388 @@
+ *  the following conditions are met:
+ *
+ *  Redistributions of source code must retain the above copyright notice, this list
+ *  of conditions and the following disclaimer.
+ *
+ *  Redistributions in binary form must reproduce the above copyright notice, this
+ *  list of conditions and the following disclaimer in the documentation and/or
+ *  other materials provided with the distribution.
+ *
+ *  Neither the name of FTC Teams 25/5218 nor the names of their contributors may be used to
+ *  endorse or promote products derived from this software without specific prior
+ *  written permission.
+ *
+ *  NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE GRANTED BY THIS
+ *  LICENSE. THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  AS IS AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ *  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESSFOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
+ *  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ *  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ *  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
+ *  TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ *  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package team25core;
+
+import com.acmerobotics.dashboard.config.Config;
+import com.qualcomm.robotcore.hardware.DcMotor;
+import com.qualcomm.robotcore.hardware.DcMotorSimple;
+import com.qualcomm.robotcore.util.RobotLog;
+
+import org.firstinspires.ftc.robotcore.external.Telemetry;
+
+@Config
+public class FourWheelDirectIMUDrivetrain extends DrivetrainBaseImpl implements DrivetrainWithIMU {
+
+    DcMotor rearLeft;
+    DcMotor rearRight;
+    DcMotor frontLeft;
+    DcMotor frontRight;
+
+    double multiplier;
+    boolean doStrafeReverse = false;
+
+    Telemetry.Item targetHeadingTlm;
+    Telemetry.Item currentYawTlm;
+    Telemetry.Item yawErrorTlm;
+
+    // FIXME proportional_constant is 0.8 on Mayfield's carpeting
+    //  try 0.9 or 0.7 to see what is better
+    // took out final and made it public so that we can see the constant in the FTC dashboard
+    // started at 0.88 (1/20)
+    public static double PROPORTIONAL_CONSTANT = 0.5; // try a little larger but lower than 0.90
+    // FIXME Temporarily set the derivative constant to 0
+    //  (I put 0.0001 initially) but first make sure
+    //  proportional constant so that it makes sure it is strafing
+    //  straight
+    public static double DERIVATIVE_CONSTANT = 0;
+
+    // this the value for the desired yaw or heading
+    // if you want to go straight this value is zero
+    private double targetYaw;
+
+    private double currentYaw;
+
+    private double currentYawRate;
+
+
```

Comment:
Remove all extraneous newlines.  Please scan the entire diff for this.  I won't comment for each instance.